File: frontend/stack-underflow_angular/src/app/services/auth.service.ts
import { Injectable, signal } from '@angular/core';
import type { User } from '../types';
import { DataStore } from '../store/data.store';

@Injectable({
  providedIn: 'root',
})
export class AuthService {
  private dataStore = new DataStore();

  userSignal = signal<User | null>(this.dataStore.currentUser);
  private initialUser = this.dataStore.currentUser;

  get user(): User | null {
    return this.userSignal();
  }

  get isAuthenticated(): boolean {
    return this.userSignal() !== null;
  }

  login(username: string, password: string): boolean {
    const user = this.dataStore.login(username, password);
    if (user) {
      this.userSignal.set(user);
      return true;
    }
    return false;
  }

  signup(username: string, password: string): boolean {
    const user = this.dataStore.signup(username, password);
    if (user) {
      this.userSignal.set(user);
      return true;
    }
    return false;
  }

  logout(): void {
    this.dataStore.logout();
    this.userSignal.set(null);
  }
}

File: frontend/stack_underflow_flutter/lib/api/auth_service.dart
import 'package:stackunderflow/api/validators.dart';
import 'package:stackunderflow/models/user.dart';
import 'package:stackunderflow/store/data_store.dart';

class AuthService {
  static final AuthService _instance = AuthService._internal();
  factory AuthService() => _instance;
  AuthService._internal();

  final DataStore _dataStore = dataStore;

  Future<User> login({
    required String username,
    required String password,
  }) async {
    await Future.delayed(const Duration(milliseconds: 200));

    validateUsername(username);
    validatePassword(password);

    _dataStore.login(username.trim(), password.trim());

    return User(
      id: 'user_${DateTime.now().millisecondsSinceEpoch}',
      username: username.trim(),
    );
  }

  Future<User> signup({
    required String username,
    required String password,
  }) async {
    await Future.delayed(const Duration(milliseconds: 200));

    validateUsername(username);
    validatePassword(password);

    _dataStore.signup(username.trim(), password.trim());

    return User(
      id: 'user_${DateTime.now().millisecondsSinceEpoch}',
      username: username.trim(),
    );
  }

  Future<void> logout() async {
    await Future.delayed(const Duration(milliseconds: 100));
    _dataStore.logout();
  }

  Future<User?> getCurrentUser() async {
    await Future.delayed(const Duration(milliseconds: 100));
    return _dataStore.getCurrentUser();
  }
}

final authService = AuthService();
